<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Cell Ranger</title>
    <meta charset="utf-8" />
    <meta name="author" content=".link-title[Course Website]" />
    <meta name="author" content=".link-title[brc@rockefeller.edu]" />
    <script src="Session1_files/header-attrs-2.26/header-attrs.js"></script>
    <link rel="stylesheet" href="xaringan-themer.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

.title[
# Cell Ranger
]
.subtitle[
## Bioinformatics Resource Center, Rockefeller University
]
.author[
### .link-title[<a href="https://rockefelleruniversity.github.io/CellRanger/">Course Website</a>]
]
.author[
### .link-title[<a href="mailto:brc@rockefeller.edu" class="email">brc@rockefeller.edu</a>]
]

---








## Overview

- .link[[Course Home Page](http://rockefelleruniversity.github.io/CellRanger/)]
- .link[[Running Cell Ranger](https://rockefelleruniversity.github.io/CellRanger/r_course/presentations/singlepage/Session1.html#Cell-Ranger)]
- .link[[Output Files](https://rockefelleruniversity.github.io/CellRanger/r_course/presentations/singlepage/Session1.html#Output-files)]
- .link[[Web Summary](https://rockefelleruniversity.github.io/CellRanger/r_course/presentations/singlepage/Session1.html#Web-Summary-QC)]

---
class: inverse, center, middle

# Set Up

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---

## Materials

Links to material and slides for this course can be found on github.

* .link[[Cell Ranger](https://rockefelleruniversity.github.io/CellRanger/)]

Or can be downloaded as a zip archive from here.

* .link[[Download zip](https://github.com/rockefelleruniversity/CellRanger/zipball/master)]

---
## Course materials

Once the zip file in unarchived. All presentations as HTML slides and pages will be available in the directories underneath.

* **r_course/presentations/slides/**
Presentations as an HTML slide show.
* **r_course/presentations/singlepage/** 
Presentations as an HTML single page.


---
class: inverse, center, middle

# Cell Ranger

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---


## A scRNAseq workflow

scRNAseq is a very variable process and each dataset has unique QC problems. Though a simple workflow often works, we often need many more tools in our toolbox to handle more complex data sets. Deciding what is appropriate often depends on the data set and its QC metrics. 

![overview](./imgs/advancedworkflow.png)

---
class: inverse, center, middle

# Cell Ranger

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---

## Cell Ranger

.pull-left[
Cell Ranger is a suite of tools for single cell processing and analysis available from 10X Genomics. It performs key processing steps i.e. mapping, and is the first chance to delve into your data sets QC. 

In this session we will give a brief overview of running this tool and then dive deeper into interpreting the outputs.
]
.pull-right[
![](./imgs/CellRangerPage.png)
]

---
## Do I need to run this?

- Often genomics centers will run it for you (i.e. here at Rockefeller)

- Why run Cell Ranger yourself?:
  * You want to integrate data sets, but they were processed by different versions of Cell Ranger
  * You may want to change Cell Ranger parameters i.e. export BAMs, or force an expected cell number.
  * Most commonly you want to reanalyze a published data set

---
## Cell Ranger Download

.pull-left[

- Cell Ranger is available from the .link[[10x genomics website](https://support.10xgenomics.com/single-cell-gene-expression/software/downloads/latest)].

- Also available are pre-baked references for Human and Mouse genomes (GRCh37/38 and GRCm37)

]
.pull-right[![](./imgs/CellRangerDownload.png)
]

---
## Cell Ranger

- Cell Ranger only runs on linux machines (CentOS/RedHat 7.0+ and Ubuntu 14.04+).
- Typically, due to memory requirements, users run Cell Ranger on a remote server and not their own machines.
- To download Cell Ranger and the required reference on remote server, we typically use the *wget* command (these commands are all on the [Download](https://support.10xgenomics.com/single-cell-gene-expression/software/downloads/latest) page).

[This will all be in terminal on the server you are using].


```shell
wget -O cellranger-8.0.0.tar.gz "https://cf.10xgenomics.com/releases/cell-exp/cellranger-8.0.0.tar.gz?Expires=1711772964&amp;Key-Pair-Id=APKAI7S6A5RYOXBWRPDA&amp;Signature=muvzcbqxba6d-blyYS02MVfLlzwZk6iZNQWXdaoCLnl7owW2nEN-IHwSPwdNoYl-6Xia7rr0S1sLCUQTsekGm2pQKcd0kqK~ndHK0DM7SwSVpXLlRvBV5pXt~EIlsxATVBKVeQLnUy698N-WnRlT~ahjlU-nMdpomX9-lOkF~w8gbgHBdtPXunTWfW87sSJLpHMDVENSF7TFJsXERDwDnsXyQLCuEhfGTCOnupkaATlLEr9kaeCStePKkwGyqgi1m8Ua02NNGHWPIJ6I1mDt695wo~dgptpJF4SDNRTyE-TuXrHfIqRjZB60zhWRJczFo2kpL7FCKwliE-vJ6djcSw__"
```

Download reference genome for Cell Ranger i.e. Human genome (GRCh38)

```shell
wget "https://cf.10xgenomics.com/supp/cell-exp/refdata-gex-GRCh38-2024-A.tar.gz"
```


---
## Cell Ranger set-up

- Having downloaded the software and references, we can then unpack them.

Unpack software and references.


```shell
tar -xzvf cellranger-8.0.0.tar.gz
tar -xzvf refdata-gex-GRCh38-2020-A.tar.gz
```

- Finally we can add the Cell Ranger directory to our PATH.


```shell
export PATH=/PATH_TO_CELLRANGER_DIRECTORY/cellranger-8.0.0:$PATH
```

---
## Running Cell Ranger count

Now we have the downloaded Cell Ranger software and required pre-build reference for Human (GRCh38) we can start the generation of count data from scRNA-seq/snRNA-seq fastQ data.

Typically FastQ files for your scRNA run will have been generated using the **Cell Ranger mkfastq** toolset to produce a directory a FastQ files. 

We can now use Cell Ranger count command with our reference and fastQ files to generate our count matrix and associated files.

If you are analyzing single nuclei RNA-seq data remember to set the **--include-introns** flag.


```shell
cellranger count --id=my_run_name \
   --fastqs=PATH_TO_FASTQ_DIRECTORY \
   --transcriptome=/PATH_TO_CELLRANGER_DIRECTORY/refdata-gex-GRCh38-2020-A
```

---
## Working with custom genomes

If you are working with a genome which is not Human and/or mouse you will need to find another source for your Cell Ranger reference.

- Luckily many references are pre-built by other consortiums.
- We can build our own references using other tools in Cell Ranger using the *mkgtf* and *mkref* functions.
- For this you just need a FASTA file (DNA sequence) and a GTF file (Gene Annotation) for your reference.

- **Importantly for Cell Ranger Count, only features labelled as exon in your GTF (_column 3_) will be considered for counting signal in genes**
- **Many genomes label mitochondrial genes with _CDS_ and not _exon_ so these must be updated**


---

class: inverse, center, middle

# Output files

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---


## Outputs

Having completed the Cell Ranger count step, the user will have created a folder named as set by the **--id** flag for the count command.

Within this folder will be the **outs/** directory containing all the outputs generated from Cell Ranger count.

![](./imgs/outputs.png)

---
## Count Matrices 

The count matrices to be used for further analysis are stored in both MEX and HDF5 formats within the output directories.

The filtered matrix only contains detected, cell-associated barcodes whereas the raw contains all barcodes (background and cell-associated).

**MEX format**
- filtered_feature_bc_matrix
- raw_feature_bc_matrix

**HDF5 format**
- filtered_feature_bc_matrix.h5
- raw_feature_bc_matrix.h5

---
## BAM files 

.pull-left[
The outs directory may also contain a BAM file of alignments for all barcodes against the reference _(possorted_genome_bam.bam)_ as well as an associated BAI index file _(possorted_genome_bam.bam.bai)_. This is no longer a default output.

This BAM file is sometimes used in downstream analysis such as scSplit/Velocyto as well as for the generation of signal graphs such as bigWigs.
]
.pull-right[
 &lt;div align="center"&gt;
&lt;img src="imgs/sam1.png" alt="igv" height="200" width="100"&gt;
&lt;/div&gt;
 &lt;div align="center"&gt;
&lt;img src="imgs/sam2.png" alt="igv" height="150" width="450"&gt;
&lt;/div&gt;

]


---
## Cloupe files 

Cell Ranger also outputs files for visualization within the .link[[10X Loupe browser software](https://www.10xgenomics.com/support/software/loupe-browser/latest)] - _cloupe.cloupe_.

This allows for the visualization of scRNA-seq/snRNA-seq as a t-sne/umap with the ability to overlay metrics of QC and gene expression onto the cells in real time

 &lt;div align="center"&gt;
&lt;img src="imgs/cloupe.png" alt="igv" height="350" width="400"&gt;
&lt;/div&gt;

---

class: inverse, center, middle

# Web Summary QC

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---


## QC outputs

Assessment of the overall quality of a scRNA-seq/snRNA-seq experiment after Cell Ranger can give our first chance to dig into the quality of your dataset and gain insight any issues we might face in data analysis.

---

## QC is essential

.pull-left[
There are many potential issues which can arise in scRNA-seq/snRNA-seq data including -

- Empty droplets
- Low quality cells (dead or dying)
- Ambient RNA contamination
- Doublets
]

.pull-right[
 &lt;div align="center"&gt;
&lt;img src="imgs/bad_cells.png" alt="igv" height="460" width="310"&gt;
&lt;caption&gt;Huemos et al. (2023)&lt;/caption&gt;
&lt;/div&gt;
]

---
## Metrics and Web Summary

Cell Ranger will also output summaries of useful metrics as a text file _(metrics_summary.csv)_ and as a intuitive web-page.

Metrics include

- Counts/UMIs per cell.
- Number of cells detected.
- Alignment quality.
- Distribution of reads in genomic features.
- Sequencing saturation
- t-sne/UMAP with default clustering.



---


## Web Summary overview

.pull-left[
The web summary html file contains an interactive report describing the most essential QC for your single cell experiment as well as initial clustering and dimension reduction for your data.

The web summary also contains useful information on the input files and the versions used in this analysis for reproducibility.
]

.pull-right[
![](./imgs/websum1.png)
]

---
## Sample panel

.pull-left[
The first thing we can review is the **Sample** information panel. **(Bottom Right)**
As most people do not run Cell Ranger themselves this is important to check it matches expectations.

- Sample ID - Sample name (Assigned in **cellranger count**).
- Chemistry - The 10x chemistry used.
- Include introns - Whether counting was run to include intron counts (typical for single neuron RNA-seq).
- Reference Path and Transcriptome - References used in analysis.
- Pipeline Version - Version of Cell Ranger used.

]

.pull-right[
![](./imgs/websum4.png)
]

---
## Sequencing panel

.pull-left[
The **Sequencing** panel highlights information on the quality of the Illumina sequencing. **Top Right**.

- Number of reads - Total number of paired reads in library.
- Valid Barcodes - Proportion of barcodes matching barcodes in whitelist (~ 740 thousand).
- Valid UMIs - Total number of UMIs that are not all one base and contain no unknown bases.
- Sequencing saturation - Unique valid barcode/UMI versus all valid barcode/UMI
- Q30 scores - Assessment of sequencing qualities for barcode/UMI/RNA Reads.

]

.pull-right[
![](./imgs/websum2.png)
]


---
## Sequencing panel

Key Metrics we look for:

  * **Q30 Bases in RNA Read &gt; 65% (usually &gt; 80%)**
		+ Reflects the sequencing quality
		+ Need to check with sequencing service supplier
		
	* **Sequencing Saturation &gt; 40% (usually range 20% ~ 80%)**
		+ Reflects the complexity of libraries
		+ Consider reconstructing library if too low

---
## Mapping panel

.pull-left[
The **Mapping** panel highlights information on the mapping of reads to the reference genome and transcriptome. **Bottom Left**

- Reads Mapped to Genome - Total number of reads mapping to the genome
- Reads Mapped Confidently to Genome - Reads mapping uniquely to the genome
- Reads Mapped Confidently to Exonic/Intronic - Reads mapping uniquely to the exons or introns
- Reads Mapped Confidently to Transcriptome - Reads mapping to a unique gene (and consistent with slice junctions).

]
.pull-right[
![](./imgs/websum3.png)
]

---
## Mapping panel

Key Metrics we look for:

  * **Mapped to Genome &gt; 60% (usually range 50% ~ 90%)**
		+ Mapping rate to reference genome
		+ Check reference genome version if too low

	* **Reads Mapped Confidently to Transcriptome &gt; 30% (usually &gt; 60%)**
		+ Reflection of annotation to transcriptome
		+ Check annotation if too low

---
## Cells panel

.pull-left[
The **Cells** panel highlights some of the most important information in the report: the total number of cells captured and the distribution of counts across cells and genes. **Top Right**

- Estimated Number of Cells - Total number of barcodes associated to at least one cell.
- Fraction Reads in Cells - Fraction of reads from valid barcode, associated to a cell and mapped to transcriptome. 
- Median Reads per Cell - Median number of transcriptome reads within cell associated barcodes
- Median Genes per Cell - - Median number of genes detected (at least 1 count) per cell associated barcodes.


]
.pull-right[
![](./imgs/websum6.png)
]


---
## Cells panel

Key Metrics we look for:

  * **Fraction Reads in Cells &gt; 70% (usually &gt; 85%)**
		+ Reflects the ambient RNA contamination
		+ Consider correcting for ambient RNA if &lt; 90%

  * **Median reads per cell &gt; 20,000/cell and estimated number of cells 500 - 10,000**
	  + May be caused by the failure of cell identification if the values were not in normal range
		+ Need to check knee plot and re-evaluate cell number

---
## The Knee plot

.pull-left[

The Cell panel also includes an interactive knee plot.

The knee plot shows:- 

- On the x-axis, the barcodes ordered by the most frequent on the left to the least frequent on the right

- On the y-axis, the frequency of each ordered barcode.

- Highlighted in dark blue are the barcodes marked as associated to cells.


]
.pull-right[
![](./imgs/websum6.png)
]

---
## Knee plot

.pull-left[

It is apparent that barcodes labelled blue (cell-associated barcodes) do not have a cut-off just based on the UMI count.

In the newer versions of Cell Ranger a two step process is used to define cell-associated barcodes based on the EmptyDrops method (Lun et al.,2019).

- First high RNA containing cells are identified based on a UMI cut-off.
- Second, low UMI containing cells are used as a background training set to identify additional cell-associated barcodes not called in first step.

If required, a **--force-cells** flag can be used with _cellranger count_ to identify a set number of cell-associated barcodes.



]
.pull-right[
![](./imgs/websum6.png)
]

---
## Knee plot

It is important to know what version and parameters were used to run Cell Ranger. This cell calling step is continually updated and it can have a dramatic affect on your results. (Cell Ranger V4.0 just dropped last month and there's been a change in default parameters).

---
## Knee plot

The Knee plot also acts a good QC tools to investigate differing types of single cell failure.

Whereas our previous knee plot represented a good sample, differing knee plot patterns can be indicative of specific problems with the single cell protocol. We will show you some examples of these below from real data.

---
## Knee plot

.pull-left[
In this example we see no specific cliff and knee suggesting a failure in the integration of oil, beads and samples (wetting failure) or a compromised sample.
]

.pull-right[
![](./imgs/wetfailure.png)
]

---
## Knee plot

.pull-left[

If there is a clog in the machine we may see a knee plot where the overall number of samples is low.



]
.pull-right[
![](./imgs/lowbarcode.png)
]

---
## Knee plot

.pull-left[

There may be occasions where we see two sets of cliff-and-knees in our knee plot.

This could be indicative of a heterogenous sample where we have two populations of cells with differing overall RNA levels.

Knee plots should be interpreted in the context of the biology under investigation.


]
.pull-right[
![](./imgs/hetero.png)
]

---
## Analysis page

.pull-left[
The web-summary also contains an analysis page where default dimension reduction, clustering and differential expressions between clusters has been performed.

Additionally the analysis page contains information on sequencing saturation and gene per cell vs reads per cell.

]

.pull-right[
![](./imgs/anal1.png)
]

---
## t-sne and clustering

.pull-left[
The t-sne plot shows the distribution and similarity within your data.

- Review for high and low UMI cells driving t-sne structure and/or clustering.
- Expected separation between and structure across clusters may be observed within the t-sne plot.
- Identify expected clusters based on expression of marker genes

]

.pull-right[
![](./imgs/anal2.png)
]

---
## Sequence and Gene saturation

.pull-left[
The sequence saturation and Median genes per cell plots show these calculations (as show on summary page) over successive downsampling of the data. 

By reviewing the curve of the down sampled metrics we can assess whether we are approaching saturation for either of these metrics.

]

.pull-right[
 &lt;div align="center"&gt;
&lt;img src="imgs/anal3.png" alt="igv" height="200" width="200"&gt;
&lt;/div&gt;
 &lt;div align="center"&gt;
&lt;img src="imgs/anal4.png" alt="igv" height="200" width="200"&gt;
&lt;/div&gt;
]


---
## How this manifests going forward?

Could show examples of high MT specifics of non-specific clustering or MT clustering. 


---
## Next steps

Fingers crossed there's no QC issues. 

Often at this step we wouldn't make any decisions unless there is a clear complete failure. This is an important first step in setting expectations/preparing for what you may need to do for the dataset.


    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
// add `data-at-shortcutkeys` attribute to <body> to resolve conflicts with JAWS
// screen reader (see PR #262)
(function(d) {
  let res = {};
  d.querySelectorAll('.remark-help-content table tr').forEach(tr => {
    const t = tr.querySelector('td:nth-child(2)').innerText;
    tr.querySelectorAll('td:first-child .key').forEach(key => {
      const k = key.innerText;
      if (/^[a-z]$/.test(k)) res[k] = t;  // must be a single letter (key)
    });
  });
  d.body.setAttribute('data-at-shortcutkeys', JSON.stringify(res));
})(document);
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
